src/acl.c

```diff
@@ -2469,18 +2469,19 @@ void addACLLogEntry(client *c, int reason, int context, int argpos, sds username
     }
 }
 
-const char* getAclErrorMessage(int acl_res) {
-    /* Notice that a variant of this code also exists on aclCommand so
-     * it also need to be updated on changed. */
+sds getAclErrorMessage(int acl_res, user *user, struct redisCommand *cmd, robj **argv, int idx) {
     switch (acl_res) {
     case ACL_DENIED_CMD:
-        return "can't run this command or subcommand";
+        return sdscatfmt(sdsempty(), "User %s has no permissions to run "
+                                     "the '%s' command or a subcommand", user->name, cmd->fullname);
     case ACL_DENIED_KEY:
-        return "can't access at least one of the keys mentioned in the command arguments";
+        return sdscatfmt(sdsempty(), "User %s has no permissions to access "
```

Comment:
I'm not sure we want to specify the exact key/channel - there can be multiple keys with no access, but the above idx is of the last one (see `ACLCheckAllUserCommandPerm()`)

Reply:
Anyway, if we decide to change the error msgs, we should update https://github.com/redis/redis-doc/blob/c1bf7b53d7f6f75e25c63379776e40345db912b5/docs/manual/security/acl.md too

Reply:
I'd note that the syntax I used was taken from acl dry run (i.e. one of the types). we can go with it being less specific, but that would then be a degrade in functionality in acl dry run.

i.e. if the inexact syntax is sufficient for acl dry run, I'd argue its good enough in general, but open to others disagreeing.

Reply:
@MeirShpilraien - what do you think?

Reply:
@MeirShpilraien - do you think we need an explicit msg in dry run mode, or can we stay with the general one (fewer details, but in case of more than multiple keys/channel - more correct)?

Reply:
I am not concerned about the generality of the error message. We are reporting basically the same thing as `ACL LOG`, which is basically the "first" key or channel that is invalid. 

I am a little bit more concerned about the fact that some security minded folks might point out that we are allowing people to discover all permission about a user, which is typically an anti-pattern. `ACL DRYRUN` should have a more comprehensive permission, since it's intended to be an admin command used for testing. I am OKAY with the more detailed error messages, as are outlined in this PR. @yossigo Do you have thoughts?

Reply:
my 2c, one does a 

`get x`

what does telling us "cmd failed" vs "cmd get failed" "hide from us" or "acl failed on key" vs "acl failed on key x".  In practice, one can determine cmd and key failures just as easily without us being clear to the user.

with that said, I'm willing to be told by security people why I'm being naive and wrong :)

Reply:
@fadidahanna I do not have a strong opinion on this one, both look fine to me, will let you choose...

Reply:
👍 so let's just wait for @yossigo response, and make sure we don't have any security issues here.

Reply:
I agree with @sjpotter that the more detailed version doesn't expose information that the user doesn't already have, but:
1. This may change in the future and go unnoticed.
3. From a security point of view, it doesn't look good and may require repeating this explanation.

I propose to unify the error message generation but maintain a flag that indicates whether we want an abbreviated or detailed message.

Reply:
are we viewing this as global switch?  i.e. a redis config option that says to use the terser or more verbose strings?

Reply:
My understanding of the comment was that we add a boolean flag to the function that indicates if we want a "verbose" or "simple" error message. Which I would be okay with.

Reply:
so lets compare to what exists now we have a "non verbose" for rm_call and scripts (relative to acl dryrun), but we have a mix for actuall processCall

https://github.com/redis/redis/blob/unstable/src/server.c#L3724

i.e. it always shows command that errored, but not key/channel.  if we want verbose to only impact key/channel, that's fine, but I guess I don't see the point if we're printing the command name everywhere.